File: 2375-minimum-obstacle-removal-to-reach-corner/minimum-obstacle-removal-to-reach-corner.java
class Solution {
    //deque
    public int minimumObstacles(int[][] grid) {
        int ROW = grid.length;
        int COL = grid[0].length;

        Deque<List<Integer>> pq = new LinkedList<>();
        HashSet<String> visited = new HashSet<String>();
        pq.add(new ArrayList<Integer>(Arrays.asList(0,0,0)));
        visited.add(String.valueOf(0) + '_' + String.valueOf(0));
        int directions[][] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        while (pq.size() > 0){
            List<Integer> temp = pq.poll();
            if (temp.get(1) == ROW - 1 && temp.get(2) == COL - 1){
                return temp.get(0);
            }
            for (int[] direction: directions){
                int x = temp.get(1) + direction[0];
                int y = temp.get(2) + direction[1];
                String key = String.valueOf(x) + '_' + String.valueOf(y);
                if (0 > x || 0 > y || x == ROW || y == COL || visited.contains(key))
                    continue;

                if (grid[x][y] > 0){
                List<Integer> t = new ArrayList<Integer>(Arrays.asList(1 + temp.get(0), x, y));
                pq.addLast(t);
                }
                else{
                List<Integer> t = new ArrayList<Integer>(Arrays.asList(temp.get(0), x, y));
                pq.addFirst(t);
                }
                
                visited.add(key);
            }
        }
        return 0; 
    }
}
    // priority queue
    // public int minimumObstacles(int[][] grid) {
    //     int ROW = grid.length;
    //     int COL = grid[0].length;

    //     PriorityQueue<List<Integer>> pq = new PriorityQueue<>((x, y) -> Integer.compare(x.get(0),y.get(0)));
    //     HashSet<String> visited = new HashSet<String>();
    //     pq.add(new ArrayList<Integer>(Arrays.asList(0,0,0)));
    //     visited.add(String.valueOf(0) + '_' + String.valueOf(0));
    //     int directions[][] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    //     while (pq.size() > 0){
    //         List<Integer> temp = pq.poll();
    //         if (temp.get(1) == ROW - 1 && temp.get(2) == COL - 1){
    //             return temp.get(0);
    //         }
    //         for (int[] direction: directions){
    //             int x = temp.get(1) + direction[0];
    //             int y = temp.get(2) + direction[1];
    //             String key = String.valueOf(x) + '_' + String.valueOf(y);
    //             if (0 > x || 0 > y || x == ROW || y == COL || visited.contains(key))
    //                 continue;
    //             List<Integer> t = new ArrayList<Integer>(Arrays.asList(grid[x][y] + temp.get(0), x, y));
    //             pq.add(t);
    //             visited.add(key);
    //         }
    //     }
    //     return 0; 
    // }
//}

File: 2375-minimum-obstacle-removal-to-reach-corner/minimum-obstacle-removal-to-reach-corner.cpp
class Solution {

public:
    int minimumObstacles(vector<vector<int>>& grid) {
        int ROW = grid.size();
        int COL = grid[0].size();

        deque<vector<int>> q;
        unordered_set<string> visited;
        q.push_back(vector<int>(3, 0));
        visited.insert(to_string(0) + '_' + to_string(0));
        vector<vector<int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        while (q.size() > 0){
            vector<int> temp = q.front();
            q.pop_front();
            if (temp[1] == ROW - 1 && temp[2] == COL - 1){
                return temp[0];
            }
            for (vector<int>& direction: directions){
                int x = temp[1] + direction[0];
                int y = temp[2] + direction[1];
                string key = to_string(x) + '_' + to_string(y);
                if (0 > x || 0 > y || x == ROW || y == COL || visited.contains(key))
                    continue;
                
                if (grid[x][y] > 0){
                    vector<int> t = {1 + temp[0], x, y};
                    q.push_back(t);
                }
                else{
                    vector<int> t = {temp[0], x, y};
                    q.push_front(t);
                }
                visited.insert(key);
            }
        }
        return 0; 
    }
    // heap
    // struct Compare{
    //     bool operator()(vector<int>& v1, vector<int>& v2){
    //         return v1 > v2;
    //     }
    // };
    // int minimumObstacles(vector<vector<int>>& grid) {
    //     int ROW = grid.size();
    //     int COL = grid[0].size();

    //     priority_queue<vector<int>, vector<vector<int>>, Compare> pq;
    //     unordered_set<string> visited;
    //     pq.push(vector<int>(3, 0));
    //     visited.insert(to_string(0) + '_' + to_string(0));
    //     vector<vector<int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    //     while (pq.size() > 0){
    //         vector<int> temp = pq.top();
    //         pq.pop();
    //         if (temp[1] == ROW - 1 && temp[2] == COL - 1){
    //             return temp[0];
    //         }
    //         for (vector<int> direction: directions){
    //             int x = temp[1] + direction[0];
    //             int y = temp[2] + direction[1];
    //             string key = to_string(x) + '_' + to_string(y);
    //             if (0 > x || 0 > y || x == ROW || y == COL || visited.contains(key))
    //                 continue;
    //             vector<int> t = {grid[x][y] + temp[0], x, y};
    //             pq.push(t);
    //             visited.insert(key);
    //         }
    //     }
    //     return 0; 
    // }
};